Update: Add getNodeByRangeIndex method (refs #1721)

btmills opened the pull request:
This spawned from the [discussion in #1721](https://github.com/eslint/eslint/pull/1721#issuecomment-71755084) about a possible way to prevent the changes to no-multi-spaces from making it and key-spacing mutually exclusive.

According to my theory, one could except nodes of type `Property` from no-multi-spaces. Then in the following example, it would check the double space between the property's key and value using `getNodeByRangeIndex`, and, finding that the whitespace belongs to the property, obey the exception and disregard the error.

```js
({
    a:  b
})
```

I thought about adding this as a comment on that discussion, but then I decided a PR would be a whole lot easier to understand.

gcochard commented:
LGTM

nzakas commented:
I think the question here is how this affects performance and if the tradeoff is worthwhile. Do you want to try building off of my PR and seeing what happens?

btmills commented:
The format I'm using for exceptions is simple:

```js
// "no-multi-spaces": [2, { "exceptions": ["Property"] }]
var obj = {
    foo:  42
};
```

Your PR, unmodified, directly from remote branch `issue1603`:

```
CPU Speed is 2200 with multiplier 7500000
Performance Run #1:  1670.4432769999999ms
Performance Run #2:  1665.231597ms
Performance Run #3:  1662.611402ms
Performance Run #4:  1681.119182ms
Performance Run #5:  1699.139068ms
Performance budget ok:  1670.4432769999999ms (limit: 3409.090909090909ms)
```

Your PR, modified to check for excepted node types, but no node types excepted:

```
CPU Speed is 2200 with multiplier 7500000
Performance Run #1:  1677.940758ms
Performance Run #2:  1664.710095ms
Performance Run #3:  1668.856047ms
Performance Run #4:  1672.700954ms
Performance Run #5:  1690.784838ms
Performance budget ok:  1672.700954ms (limit: 3409.090909090909ms)
```

Your PR, modified to check for excepted node types, and `Property` nodes excepted:

```
CPU Speed is 2200 with multiplier 7500000
Performance Run #1:  1668.243735ms
Performance Run #2:  1673.431812ms
Performance Run #3:  1777.5240840000001ms
Performance Run #4:  1689.695841ms
Performance Run #5:  1667.02142ms
Performance budget ok:  1673.431812ms (limit: 3409.090909090909ms)
```

The perf impact looks to be statistically insignificant in light of the variance between the runs themselves. I think this would be worth it. If you agree, how would you want to structure the commits/branches/PRs to get everything merged? Perhaps merge #1721, I'll rebase this so it can be merged (I want to keep `get(Token|Node)ByRangeIndex` next to each other in the source, and automatically merging won't do that), then I'll create an issue and a PR to add exceptions to the rule?

nzakas commented:
That sounds like an excellent plan. Let's do it.

nzakas commented:
@btmills do we still need this?

btmills commented:
@nzakas The fix for #1755 will depend on this method. If you want me to roll this into that upcoming PR I can, but since it's already been reviewed, I've rebased this, and it's ready to go.

nzakas commented:
Cool!
